gnd/src/main/java/com/google/android/gnd/ui/editobservation/EditObservationViewModel.java

```diff
@@ -303,21 +293,13 @@ private void onObservationLoaded(Observation observation) {
   }
 
   private Single<Event<SaveResult>> save() {
-    savingProgressVisibility.setValue(View.VISIBLE);
-    ObservationMutation observationMutation =
-        ObservationMutation.builder()
-            .setType(isNew ? ObservationMutation.Type.CREATE : ObservationMutation.Type.UPDATE)
-            .setProjectId(originalObservation.getProject().getId())
-            .setFeatureId(originalObservation.getFeature().getId())
-            .setLayerId(originalObservation.getFeature().getLayer().getId())
-            .setObservationId(originalObservation.getId())
-            .setFormId(originalObservation.getForm().getId())
-            .setResponseDeltas(getResponseDeltas())
-            .setClientTimestamp(new Date())
-            .setUserId(authManager.getCurrentUser().getId())
-            .build();
+    if (originalObservation == null) {
+      return Single.error(new IllegalStateException("Observation is null"));
+    }
+
     return observationRepository
-        .applyAndEnqueue(observationMutation)
+        .addObservationMutation(originalObservation, getResponseDeltas(), isNew)
+        .doOnSubscribe(disposable -> savingProgressVisibility.setValue(View.VISIBLE))
```

Comment:
Just an observation, the viewmodel's `saveProgressVisibility` property could be refactored to `isSaving` and the visibility logic could then be handled in the layout XML using databinding e.g. `android:visibility="@{viewmodel.isSaving ? View.VISIBLE : View.GONE}"`

Reply:
Thanks for the comment! We are already using databinding for toggling the visibility. The only difference is, currently the binded object is an integer type whose value is equal to one of the visibility values. And in this scenario, we would use a boolean type instead with the suggested refactor. Any particular reason for using boolean over integer type?

Reply:
Yeah, it's really only a minor thing. The reason for using boolean over integer is that it represents the 2 saving states: "saving" and "not saving". 

Currently the `savingProgressVisibility` integer maps to the 3 possible visibility states: visible, invisible and gone. We don't seem to use "invisible" so refactoring to `isSaving.setValue(true)` makes the code (albeit only slightly) more readable.

Reply:
I did the refactor on `ObservationListViewModel` - see what you think :) https://github.com/shobhitagarwal1612/ground-android/pull/1/files

Reply:
I like this better. Thanks for the suggestion! 
Would you like to open a PR with the proposed changes? If not, then I'd be more than happy to do it on your behalf. 😄

Reply:
I'm happy to do it. Will do shortly.

Reply:
https://github.com/google/ground-android/pull/529